Adds procedures for configuring private storage endpoints with Azure (#69024)
openshift/openshift-docs: modules/configuring-private-storage-endpoint-azure-user-provided-vnet-subnet.adoc

// Module included in the following assemblies:
//
// * post_installation_configuration/configuring-private-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="configuring-private-storage-endpoint-azure-user-provided-vnet-subnet"]
= Configuring a private storage endpoint on Azure with user-provided VNet and subnet names

Use the following procedure to configure a storage account that has public network access disabled and is exposed behind a private endpoint on Azure.

.Prerequisites

> * You have configured an image registry to run on Azure

Reviewer: ```suggestion
* You have configured the image registry to run on Azure
```

> * You must know the VNet and subnet names used for your Azure environment.

Reviewer: I forgot to mention users may also have a `networkResourceGroupName` - this was introduced to the api after the first round of PRs merged, so it wasn't present in the initial testing steps I provided.
```suggestion
* You must know the VNet and subnet names used for your Azure environment.
* If your network was configured in a separate resource group in Azure, you must also know it's name.
```

.Procedure 

. Edit the Image Registry Operator `config` object and configure the private endpoint using your VNet and subnet names:
+
[source,terminal]
----
$ oc edit configs.imageregistry/cluster
----
+
[source,terminal]
----
# ...
spec:
  [...]
   storage:
      azure:
        [...]
        networkAccess:
          type: Internal
          internal:
            subnetName: <subnet_name>

>             vnetName: <vnet_name>

Reviewer: ```suggestion
            vnetName: <vnet_name>
            networkResourceGroupName: <network_resource_group_name>
```

Author: added.

# ...
----

. Optional. Enter the following command to confirm that the Operator has completed provisioning. This might take a few minutes. 
+
[source,terminal]
----
$ oc get configs.imageregistry/cluster -o=jsonpath="{.spec.storage.azure.privateEndpointName}" -w
----

> ////

Reviewer: Are we missing the "When redirect is enabled, pulling images from outside of the cluster is not expected to work." note here too?

Author: added

. Optional. If the registry is exposed via a route, whether the `defaultRoute` or otherwise, disable `redirect` on the Image Operator configuration by entering the following command:
+
[source,terminal]
----
$ oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"disableRedirect": true}}'
----
////

> .Verification

Reviewer: Same comment for the verification on the other procedure applies here.

// Module included in the following assemblies:
//
// * post_installation_configuration/configuring-private-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="configuring-private-storage-endpoint-azure-user-provided-vnet-subnet"]
= Configuring a private storage endpoint on Azure with user-provided VNet and subnet names

> Use the following procedure to configure a private storage endpoint on Azure with user-provided VNet and subnet names.

Reviewer: Private storage endpoint is not a thing in azure. I'm okay with introducing this term here as it does make sense. The terms used in azure are "private endpoint" and "storage account". In the code I've been using "private storage account" to refer to a storage account that has public network access disabled and is exposed behind a private endpoint.
I'll leave it to you to decide how to proceed.

.Prerequisites 

> * You have configured the Image Registry Operator for use with Azure.

Reviewer: What does this mean exactly? Is it about having the `ImageRegistry` capability enabled?

Author: I guess I'm not really sure. What are the required prerequisites beyond "You must know the VNet and subnet names used in your Azure environment." ? 

I assume that they should have configured a private Azure cluster, right?

Reviewer: With a public cluster users will still be able to use this feature (although it doesn't make as much sense).
I guess the only prerequisite is that there is a registry in their cluster? And that the storage is managed by the operator (`.spec.storage.managementState: Managed`) but but these two things are defaults on new installations so I'm not sure if it's worth mentioning 🤔

Author: Added the prerequisite: * You have configured an image registry to run on Azure

.Procedure 

. Enter the following command to obtain the VNet and subnet names that are used in the cluster:
+
[source,terminal]
----

> $ oc get machinesets -nopenshift-machine-api -ojson|jq '.items[1].spec.template.spec.providerSpec.value | (.vnet, .subnet)'

Reviewer: When users bring their own vnet and subnets they usually know their names because they configured it themselves (usually before creating the cluster, if I understand it correctly).
To me, this means that they don't need to figure out what the correct vnet and subnet names are. If they don't know their vnet and subnet names, it probably means their cluster is using the installer provisioned vnet and subnets. In this case, leaving the discovery to the operator is the best way to use this feature.

Maybe adding "know the vnet and subnet names" to the prerequisites list for this specific procedure? It might also be worth placing this procedure after the one that uses the operator to discover vnet and subnet names, since the bring your own vnet and subnet is more advanced than the operator discovery procedure. WDY?

Side note: I added this to the testing steps in the PR because it was an easy way to simulate an installation where the user brought their own vnet and subnets.

Author: @flavianmissi I switched the procedures to that the other is now first. How's this bullet point?

* You must know the VNet and subnet names used in your Azure environment.

Reviewer: I think we want to keep this step. The bullet point sound good for the "bring your own vnet/subnet" configuration!

----

. Edit the Image Registry Operator `config/instance` object and configure the private endpoint using your VNet and subnet names:
+
[source,terminal]
----
$ oc edit configs.imageregistry/cluster
----
+
[source,terminal]
----
# ...
spec:
  [...]
   storage:
      azure:
        [...]
        networkAccess:
          type: Internal
          internal:
            subnetName: <subnet_name>
            vnetName: <vnet_name>
# ...
----

> . Wait until the Image Registry Operator configures the private endpoint. Enter the following command to return the endpoint name:

Reviewer: I had this as a testing step because it helped me know when the operator was done configuring the private endpoint. End users don't need to know the private endpoint name.
Maybe we make this an optional step? Maybe something like "(Optional) Watch the private endpoint name to know when the operator completed provisioning - this may take a few minutes"?

Author: Made it optional in both procedures.

+
[source,terminal]
----
$ oc get configs.imageregistry/cluster -o=jsonpath="{.spec.storage.azure.privateEndpointName}" -w
----

> . Enter the following command to enable the default route and disable `redirect` on the Image Operator configuration:
> +
> [source,terminal]
> ----
> $ oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"defaultRoute": true, "disableRedirect": true}}'

Reviewer: Enabling the default route is not a required step to configure private endpoints.
Disabling redirect is only required when users already expose the registry (via defaultRoute or a custom route).

Enabling the route (and disabling redirect) should be located in the verification steps. 
```suggestion
. (Optional) If the registry is exposed via a route (defaultRoute or otherwise), disable `redirect` on the Image Operator configuration:
+
[source,terminal]
----
$ oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"disableRedirect": true}}'
```

Author: Fixed language in both procedures.

----

> .Verification

Reviewer: What do we usually want to achieve in the verification section?
Showing that pulls work does not prove that the storage account is private. If we want to prove that the previous configuration worked, we might want to show that pulling via default route _without disabling redirect_ won't work - this is the way to prove that the private endpoint configuration worked.
I'm not sure if it makes sense to do this here tho... WDYT?

Reviewer: I like it! left some comments there.

// Module included in the following assemblies:
//
// * post_installation_configuration/configuring-private-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="configuring-private-storage-endpoint-azure-user-provided-vnet-subnet"]
= Configuring a private storage endpoint on Azure with user-provided VNet and subnet names

Use the following procedure to configure a storage account that has public network access disabled and is exposed behind a private endpoint on Azure.

.Prerequisites

* You have configured an image registry to run on Azure
* You must know the VNet and subnet names used for your Azure environment.
* You have configured a route

.Procedure 

> . Enter the following command to obtain the VNet and subnet names that are used in the cluster:
> +
> [source,terminal]
> ----
> $ oc get machinesets -nopenshift-machine-api -ojson|jq '.items[1].spec.template.spec.providerSpec.value | (.vnet, .subnet)'
> ----

Reviewer: Users should know their vnet and subnet names for this procedure, we should not tell them to get it from machinesets (they might even get the wrong names from this command)

```suggestion

```

// Module included in the following assemblies:
//
// * post_installation_configuration/configuring-private-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="configuring-private-storage-endpoint-azure-user-provided-vnet-subnet"]

> = Configuring a private storage endpoint on Azure with user-provided VNet and subnet names

Reviewer 2: maybe it's better replace VNet with vnet

Author: Thanks. For the time being, I think I'll keep it as VNet because in our docs it's most commonly "VNet". Will talk to the lead writer to see what he thinks about changing all instances in the future.

> Use the following procedure to configure a storage account that has public network access disabled and is exposed behind a private endpoint on Azure.

Reviewer: nit:
```suggestion
Use the following procedure to configure a storage account that has public network access disabled and is exposed behind a private storage endpoint on Azure.
```

.Prerequisites

> * You have configured the image registry to run on Azure

Author: Fixed x3

* You must know the VNet and subnet names used for your Azure environment.
* If your network was configured in a separate resource group in Azure, you must also know its name.

.Procedure 

. Edit the Image Registry Operator `config` object and configure the private endpoint using your VNet and subnet names:
+
[source,terminal]
----
$ oc edit configs.imageregistry/cluster
----
+
[source,terminal]
----
# ...
spec:
  [...]
   storage:
      azure:
        [...]
        networkAccess:
          type: Internal
          internal:
            subnetName: <subnet_name>
            vnetName: <vnet_name>
            networkResourceGroupName: <network_resource_group_name>
# ...
----

> . Optional. Enter the following command to confirm that the Operator has completed provisioning. This might take a few minutes. 

Author: fixed x2

+
[source,terminal]
----
$ oc get configs.imageregistry/cluster -o=jsonpath="{.spec.storage.azure.privateEndpointName}" -w
----
+
[NOTE]
====
When redirect is enabled, pulling images from outside of the cluster will not work.
====

.Verification

> Use the following verification steps to ensure that users within the same cluster can pull images from the registry.

Reviewer 3: I don't think we normally want a lead-in sentence before steps

https://redhat-documentation.github.io/supplementary-style-guide/#lead-in-sentences

Doesn't specify verification steps, but I assume that would be the same as procedure guidelines

Same comment for other modules

. Fetch the registry service name by running the following command:
+
[source,terminal]
----
$ oc registry info --internal=true
----
+
.Example output
+
[source,terminal]
----
image-registry.openshift-image-registry.svc:5000
----

. Enter debug mode by running the following command:
+
[source,terminal]
----

> $ oc debug node/<node-name>

Reviewer 3: ```suggestion
$ oc debug node/<node_name>
```

Author: fixed x2
